src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/FSharpCoreReflectionProxy.cs

```diff
@@ -0,0 +1,250 @@
+    // Every F# application ships with a copy of FSharp.Core, however it is not available statically to System.Text.Json.
+    // The following class uses reflection to access the relevant APIs required to detect the various F# types we are looking to support.
+
+    /// <summary>
+    /// Proxy class used to access FSharp.Core metadata and reflection APIs that are not statically available to System.Text.Json.
+    /// </summary>
+    internal sealed class FSharpCoreReflectionProxy
+    {
+        /// <summary>
+        /// The various categories of F# types that System.Text.Json supports.
+        /// </summary>
+        public enum FSharpKind
+        {
+            Unrecognized,
+            Option,
+            ValueOption,
+            List,
+            Set,
+            Map,
+            Record
+        }
+
+        // Binding a struct getter method to a delegate requires that the struct parameter is passed byref.
+        public delegate TResult StructGetter<TStruct, TResult>(ref TStruct @this) where TStruct : struct;
+
+        public const string FSharpCoreUnreferencedCodeMessage = "Uses Reflection to access FSharp.Core components at runtime.";
+
+        private static object s_lockObj = new object();
+        private static FSharpCoreReflectionProxy? s_singletonInstance;
+
+        // Every type generated by the F# compiler is annotated with the CompilationMappingAttribute
+        // containing all relevant metadata required to determine its kind:
+        // https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-compilationmappingattribute.html#SourceConstructFlags
+        private const string CompilationMappingAttributeTypeName = "Microsoft.FSharp.Core.CompilationMappingAttribute";
+        private readonly Type _compilationMappingAttributeType;
+        private readonly MethodInfo? _sourceConstructFlagsGetter;
+
+        private readonly Type? _fsharpOptionType;
+        private readonly Type? _fsharpValueOptionType;
+        private readonly Type? _fsharpListType;
+        private readonly Type? _fsharpSetType;
+        private readonly Type? _fsharpMapType;
+
+        private readonly MethodInfo? _fsharpListCtor;
+        private readonly MethodInfo? _fsharpSetCtor;
+        private readonly MethodInfo? _fsharpMapCtor;
+
+        /// <summary>
+        /// Checks if the provided System.Type instance is emitted by the F# compiler.
+        /// If true, also initializes the proxy singleton for future by other F# types.
+        /// </summary>
+        [RequiresUnreferencedCode(FSharpCoreUnreferencedCodeMessage)]
+        public static bool IsFSharpType(Type type)
+        {
+            if (s_singletonInstance is null)
+            {
+                if (GetFSharpCoreAssembly(type) is Assembly fsharpCoreAssembly)
+                {
+                    // Type is F# type, initialize the singleton instance.
+                    lock (s_lockObj)
```

Comment:
Is this lock necessary? Can `s_singletonInstance` just be overwritten, perhaps add `volatile` to it if is is expensive to re-create on multiple threads.

Reply:
You're right that it's probably not necessary. Removed.